nni/retiarii/experiment.py

```diff
@@ -156,42 +169,79 @@ def start(self, config: RetiariiExeConfig, port: int = 8080, debug: bool = False
 
         # dispatcher must be created after pipe initialized
         # the logic to launch dispatcher in background should be refactored into dispatcher api
-        Thread(target=self._dispatcher.run).start()
+        self._dispatcher_thread = Thread(target=self._dispatcher.run)
+        self._dispatcher_thread.start()
 
         self._start_strategy()
 
+        ips = [self.config.nni_manager_ip]
+        for interfaces in psutil.net_if_addrs().values():
+            for interface in interfaces:
+                if interface.family == socket.AF_INET:
+                    ips.append(interface.address)
+        ips = [f'http://{ip}:{port}' for ip in ips if ip]
+        msg = 'Web UI URLs: ' + colorama.Fore.CYAN + ' '.join(ips)
+        _logger.info(msg)
+
         # TODO: register experiment management metadata
 
     def stop(self) -> None:
         """
         Stop background experiment.
         """
-        self._proc.kill()
-        self._pipe.close()
+        _logger.info('Stopping experiment...')
+        atexit.unregister(self.stop)
+
+        if self._proc is not None:
+            kill_command(self._proc.pid)
+        if self._pipe is not None:
+            self._pipe.close()
+        if self._dispatcher_thread is not None:
+            self._dispatcher.stopping = True
+            self._dispatcher_thread.join(timeout=1)
 
         self.port = None
         self._proc = None
         self._pipe = None
+        self._dispatcher = None
+        self._dispatcher_thread = None
 
-    def run(self, config: RetiariiExeConfig, port: int = 8080, debug: bool = False) -> str:
+    def run(self, config: RetiariiExeConfig = None, port: int = 8080, debug: bool = False) -> str:
```

Comment:
This looks identical to base class. Why overriding?

Reply:
this one is not identical, because this is `OneShotTrainers`

Reply:
`get_status` can be removed

Reply:
Copied part should be `super().run(...)`

Reply:
update, thanks

Reply:
It also applies to `start` and `stop`...

Reply:
yeah, `stop` can be removed. but `start` cannot, because dispatcher is instantiated in it.